include/supernode.h

```diff
@@ -1,38 +1,92 @@
 #pragma once
 #include <boost/optional.hpp>
+#include <fstream>
+#include <sys/mman.h>
+
 #include "sketch.h"
 
 using namespace std;
 
 typedef uint64_t Node;
 typedef std::pair<Node, Node> Edge;
 
+// #define USE_MLOCK
+
 /**
  * This interface implements the "supernode" so Boruvka can use it as a black
  * box without needing to worry about implementing l_0.
  */
 class Supernode {
-  /* collection of logn sketches to query from, since we can't query from one
-     sketch more than once */
-  vector<Sketch*> sketches;
+  static uint32_t bytes_size;
   int idx;
   int logn;
   std::mutex node_mt;
 
   FRIEND_TEST(SupernodeTestSuite, TestBatchUpdate);
   FRIEND_TEST(SupernodeTestSuite, TestConcurrency);
+  FRIEND_TEST(SupernodeTestSuite, TestSerialization);
+  FRIEND_TEST(GraphTestSuite, TestCorrectnessOfReheating);
   FRIEND_TEST(EXPR_Parallelism, N10kU100k);
 
 public:
   const uint64_t n; // for creating a copy
   const long seed; // for creating a copy
+  
+private:
+  size_t sketch_size;
+
+  /* collection of logn sketches to query from, since we can't query from one
+     sketch more than once */
+  // The sketches, off the end.
+  alignas(Sketch) char sketch_buffer[1];
+  
   /**
    * @param n     the total number of nodes in the graph.
    * @param seed  the (fixed) seed value passed to each supernode.
    */
   Supernode(uint64_t n, long seed);
+
+  /**
+   * @param n         the total number of nodes in the graph.
+   * @param seed      the (fixed) seed value passed to each supernode.
+   * @param binary_in A stream to read the file from
+   */
+  Supernode(uint64_t n, long seed, std::fstream &binary_in);
+
+public:
+  using SupernodeUniquePtr = std::unique_ptr<Supernode, std::function<void(Supernode*)>>;
+  static SupernodeUniquePtr makeSupernode(uint64_t n, long seed);
+  static SupernodeUniquePtr makeSupernode(uint64_t n, long seed, std::fstream &binary_in);
+  static Supernode* makeSupernode(void* loc, uint64_t n, long seed);
+
   ~Supernode();
 
+  inline Sketch* get_sketch(size_t i)
+  { return reinterpret_cast<Sketch*>(sketch_buffer + i * sketch_size); }
+
+  inline const Sketch* get_sketch(size_t i) const
+  { return reinterpret_cast<const Sketch*>(sketch_buffer + i * sketch_size); }
+
+  static inline void set_size(uint64_t n) 
+  { bytes_size = sizeof(Supernode) + log2(n) * Sketch::sketchSizeof() - sizeof(char);}
+
+  static inline uint32_t get_size() 
+  { return bytes_size; }
+
+  static inline void page_in(Supernode* sn) { 
+#ifdef USE_MLOCK
+    madvise(sn, bytes_size, MADV_WILLNEED); 
+#else
+    mlock(sn, bytes_size);
+#endif
+  }
+
+#ifdef USE_MLOCK 
+  static inline void page_out(Supernode* sn) 
+  { munlock(sn, bytes_size); }
+#else 
+  static inline void page_out(Supernode *) {}
+#endif
```

Comment:
remove

Reply:
Maybe, depends if it helps with insertion performance. It probably won't but I want to check.